go-controller/cmd/ovnkube/ovnkube.go

```diff
@@ -426,159 +426,174 @@ func runOvnKube(ctx context.Context, runMode *ovnkubeRunMode, ovnClientset *util
 		return ovnnode.CleanupClusterNode(runMode.identity)
 	}
 
-	stopChan := make(chan struct{})
-	wg := &sync.WaitGroup{}
-	defer func() {
-		close(stopChan)
-		wg.Wait()
-	}()
-
-	var masterWatchFactory *factory.WatchFactory
-
-	if runMode.ovnkubeController {
-		// create factory and start the controllers asked for
-		masterWatchFactory, err = factory.NewOVNKubeControllerWatchFactory(ovnClientset.GetOVNKubeControllerClientset())
-		if err != nil {
-			return err
-		}
-		defer masterWatchFactory.Shutdown()
+	watchFactory, err := newWatchFactory(runMode, ovnClientset)
+	if err != nil {
+		return fmt.Errorf("failed to initialize watch factory: %w", err)
 	}
 
+	// there might be dependencies accross components when starting so run them
```

Comment:
~so currently if my CM has start up issues that thread will return and we will try to start up the NCM next, is that desired to have NCM run if CM isn't running? (maybe answer is yes in IC and no in NONIC?)
PS: It's hard to track how this used to be before this change since that code was too much nested, just checking to ensure behaviour hasn't changed.~
ah nvm since the context is same parent context, I guess cancelling one will call Done for all?